File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.py
import sys
input = sys.stdin.readline

idx = 1
N = int(input())
l = list(map(int, input().strip().split()))

stack = []
for i in l:
    if i != idx:
        stack.append(i)
        continue

    idx += 1

    while stack and stack[-1] == idx:
        stack.pop()
        idx += 1
        
if not stack:
    print("Nice")
else:
    print("Sad")

File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        int idx = 1;
        int N = Integer.parseInt(br.readLine());
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        int[] l = new int[N];
        for (int i = 0; i < N; i++) {
            l[i] = Integer.parseInt(st.nextToken());
        }
        
        Stack<Integer> stack = new Stack<>();
        
        for (int i = 0; i < N; i++) {
            int person = l[i];
            
            if (person != idx) {
                stack.push(person);
                continue;
            }
            
            idx++;
            
            while (!stack.isEmpty() && stack.peek() == idx) {
                stack.pop();
                idx++;
            }
        }
        
        if (stack.isEmpty()) {
            System.out.println("Nice");
        } else {
            System.out.println("Sad");
        }
    }
}